### PR TITLE
Unfuck PruneDeadFields

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Reassigned to @catoverdrive

**Comment by tpoterba** [comment]
This is high-prio because it's blocking most of my other open PRs

**Comment by tpoterba** [comment]
Thanks!

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 1]
```diff
@@ -763,47 +842,130 @@ object PruneDeadFields {
...
+            BindingEnv(scan = Some(keyEnv.eval))
+          else
+            BindingEnv(agg = Some(keyEnv.eval)),
+          memoizeValueIR(aggIR, requestedType, memo)
+        )
+      case AggArrayPerElement(a, name, aggBody, isScan) =>
         val aType = a.typ.asInstanceOf[TArray]
-        val bodyDep = memoizeValueIR(aggBody,
+        val bodyEnv = memoizeValueIR(aggBody,
           requestedType.asInstanceOf[TArray].elementType,
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: how is this somehow getting a TDict as the requested type? Python tests are failing here, I believe?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: yep debugging, found some other stuff too (like no memoize rule for ArraySort, which led to free variables getting propagated up -- this wasn't a bug until I changed the prune structure)

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 2]
```diff
@@ -763,47 +842,130 @@ object PruneDeadFields {
...
         )
-      case AggArrayPerElement(a, name, aggBody, _) =>
+      case AggGroupBy(key, aggIR, isScan) =>
+        val keyEnv = memoizeValueIR(key, key.typ, memo)
+        unifyEnvs(
+          if (isScan)
+            BindingEnv(scan = Some(keyEnv.eval))
+          else
+            BindingEnv(agg = Some(keyEnv.eval)),
+          memoizeValueIR(aggIR, requestedType, memo)
```
Comment (catoverdrive) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: ah, found it. I *think* this should be
```
memoizeValueIR(aggIR, coerce[TDict](requestedType).valueType, memo)
```

**Pull request merged**